controller/controller_server/utils.py

```diff
@@ -8,43 +10,164 @@
+    try:
+        parsed_input = json.loads(raw_input)
+    except json.decoder.JSONDecodeError:
+        raise ValidationException(messages.invalid_non_json_parameter_message.format(raw_input))
+    return parsed_input
+
+
+def _is_topology_match(system_topologies, node_topologies):
+    for topologies in system_topologies:
+        logger.debug(
+            "Comparing topologies: system topologies: {},"
+            " node topologies: {}".format(topologies, node_topologies))
+        if topologies.items() <= node_topologies.items():
+            return True
+    return False
+
+
+def get_volume_topologies(request):
+    if request.HasField(config.REQUEST_ACCESSIBILITY_REQUIREMENTS_FIELD):
+        accessibility_requirements = request.accessibility_requirements
+        if accessibility_requirements.preferred:
+            topologies = accessibility_requirements.preferred[0].segments
+            logger.info("Chosen volume topologies: {}".format(topologies))
+            return topologies
+
+
+def get_system_info_by_topologies(secrets_config, node_topologies):
+    for system_id, system_info in secrets_config.items():
+        system_topologies = system_info.get(config.SECRET_SUPPORTED_TOPOLOGIES_PARAMETER)
+        if _is_topology_match(system_topologies, node_topologies):
+            return system_info, system_id
+
+
+def _get_system_info_by_system_id(secrets_config, system_id):
+    return secrets_config.get(system_id)
+
+
+def _get_system_info_from_secrets(secrets, topologies=None, system_id=None):
+    raw_secrets_config = secrets.get(config.SECRET_CONFIG_PARAMETER)
+    system_info = secrets
+    if raw_secrets_config:
+        secrets_config = _parse_raw_input(raw_input=raw_secrets_config)
+        if system_id:
+            system_info = _get_system_info_by_system_id(secrets_config=secrets_config, system_id=system_id)
+        elif topologies:
+            system_info, system_id = get_system_info_by_topologies(secrets_config=secrets_config,
+                                                                   node_topologies=topologies)
+        else:
+            raise ValidationException(messages.invalid_secrets_message)
+    return system_info, system_id
+
+
+def _get_array_connection_info_from_system_info(secrets, system_id):
     user = secrets[config.SECRET_USERNAME_PARAMETER]
     password = secrets[config.SECRET_PASSWORD_PARAMETER]
-    array_addresses = secrets[config.SECRET_ARRAY_PARAMETER].split(",")
-    return user, password, array_addresses
+    array_addresses = secrets[config.SECRET_ARRAY_PARAMETER].split(config.PARAMETERS_ARRAY_ADDRESSES_DELIMITER)
+    return ArrayConnectionInfo(array_addresses=array_addresses, user=user, password=password, system_id=system_id)
+
+
+def get_array_connection_info_from_secrets(secrets, topologies=None, system_id=None):
+    system_info, system_id = _get_system_info_from_secrets(secrets, topologies, system_id)
+    return _get_array_connection_info_from_system_info(system_info, system_id)
```

Comment:
so if there is no topology match, we get a non-indicative key error in `secrets[config.SECRET_USERNAME_PARAMETER]`?

Reply:
error

Reply:
which error?

Reply:
in [get_system_info_by_topologies ](https://github.com/IBM/ibm-block-csi-driver/blob/7d4d4442ddbc18cfe786657eabd85122356ad71d/controller/controller_server/utils.py#L52)

Reply:
great! why is it public btw?

Reply:
right, now it's private